src/main/scala/org/constellation/storage/RateLimiting.scala

```diff
@@ -0,0 +1,68 @@
+package org.constellation.storage
+
+import cats.effect.Concurrent
+import cats.implicits._
+import io.chrisdavenport.log4cats.Logger
+import org.constellation.primitives.concurrency.SingleRef
+import org.constellation.primitives.{Schema, Transaction}
+
+import scala.math.ceil
+
+class RateLimiting[F[_]: Concurrent: Logger]() {
+  private[storage] val counter: SingleRef[F, Map[Schema.Address, Int]] = SingleRef[F, Map[Schema.Address, Int]](Map())
+  private[storage] val blacklisted: StorageService[F, Int] = new StorageService()
+
+  def update(txs: List[Transaction]): F[Unit] =
+    for {
+      _ <- counter.acquire
+      grouped <- txs.groupBy(_.src).mapValues(_.size).pure[F]
+      _ <- counter.updateUnsafe(_ |+| grouped)
+
+      _ <- counter.getUnsafe.flatTap(c => Logger[F].debug(s"Update [rate-limiting]: $c"))
+      _ <- blacklist()
+      _ <- counter.release
+    } yield ()
+
+  def reset(cbHashes: List[String])(checkpointService: CheckpointService[F]): F[Unit] =
+    for {
+      _ <- counter.acquire
+      cbs <- cbHashes.traverse(checkpointService.fullData).map(_.flatten)
```

Comment:
This could be outside the lock since it's local variable

Reply:
Yes, changed